include/field.hxx

```diff
@@ -143,6 +144,36 @@ class Field {
    * Return the number of nz points
    */
   virtual int getNz() const;
+
+  virtual bool is3D() const {
+    throw BoutException("not implemented in Field base class");
```

Comment:
why not set this to false?
then it only needs to overwritten once ...

Reply:
Setting to false as you say would make sense. The version in FieldData is

    virtual bool is3D() const = 0; ///< True if variable is 3D

So the function has to be implemented in every derived class.

But I think the design here needs some thought. There are both Field and FieldData base classes; Field3D and Field2D inherit from both, while FieldPerp only inherits from Field. The code in

    Field3D toFieldAligned(const Field &f)

basically needs to know if the Field f is a Field3D, Field2D or FieldPerp. At the moment it works 'correctly' because:
- for Field3D it calls toFieldAligned
- for Field2D it casts the Field to a Field2D and then creates a Field3D by assignment from the Field2D, then calls toFieldAligned on the Field3D
- for FieldPerp throws an exception because FieldPerp does not implement is3D()
but especially for FieldPerp, this is a strange and confusing way to do things. Maybe it would be better to make 'toFieldAligned' be a method of Field/Field3D/etc.? Then it would be easy to get the right method for each type from a generic Field.